src/execution/index/art/fixed_size_allocator.cpp

```diff
@@ -0,0 +1,221 @@
+#include "duckdb/execution/index/art/fixed_size_allocator.hpp"
+
+#include "duckdb/common/allocator.hpp"
+#include "duckdb/common/exception.hpp"
+#include "duckdb/common/helper.hpp"
+
+namespace duckdb {
+
+constexpr idx_t FixedSizeAllocator::BASE[];
+constexpr uint8_t FixedSizeAllocator::SHIFT[];
+
+FixedSizeAllocator::FixedSizeAllocator(const idx_t &allocation_size)
+    : allocation_size(allocation_size), total_allocations(0) {
+
+	// calculate how many allocations fit into one buffer
+
+	idx_t bits_per_value = sizeof(validity_t) * 8;
+	idx_t curr_alloc_size = 0;
+
+	bitmask_count = 0;
+	allocations_per_buffer = 0;
+
+	while (curr_alloc_size < BUFFER_ALLOCATION_SIZE) {
+		if (!bitmask_count || (bitmask_count * bits_per_value) % allocations_per_buffer == 0) {
+			bitmask_count++;
+			curr_alloc_size += sizeof(validity_t);
+		}
+
+		auto remaining_alloc_size = BUFFER_ALLOCATION_SIZE - curr_alloc_size;
+		auto remaining_allocations = MinValue(remaining_alloc_size / allocation_size, bits_per_value);
+
+		if (remaining_allocations == 0) {
+			break;
+		}
+
+		allocations_per_buffer += remaining_allocations;
+		curr_alloc_size += remaining_allocations * allocation_size;
+	}
+
+	allocation_offset = bitmask_count * sizeof(validity_t);
+}
+
+FixedSizeAllocator::~FixedSizeAllocator() {
+	for (auto &buffer : buffers) {
+		Allocator::DefaultAllocator().FreeData(buffer.ptr, BUFFER_ALLOCATION_SIZE);
```

Comment:
Perhaps pass in the allocator as a parameter to the `FixedSizeAllocator` instead, and pass in the `BufferAllocator` from the buffer manager which includes automatic memory tracking (`BufferManager::GetBufferManager(context).GetBufferAllocator()`)